### ISSUE TITLE
Create annotations programmatically

### ISSUE THREAD

**Issue by ilpaschi** [mozilla/pdf.js#16168]
Hello there!
I'd like to create annotations (like the Ink one) programmatically via Javascript code.
Is there an API / Function that can be invoked to create an annotation on a specific page and on a specific location programmatically?

Ideally i'd need to create a rectangle.

Is there any way to do so?
Thanks!
Stefano

**Comment by Snuffleupagus** [comment]
Please see #15712

**Comment by ilpaschi** [comment]
Hello, thanks for the feedback.
I don't need to add an annotation through the "AnnotationEditorLayer". I need to mimick what happens when the source PDF has a Square annotation in it, and add a new one programmatically.

I can access the AnnotationStorage of a page, but if I push new annotations there (i.e. a copy of the original annotation from the source PDF with different coordinates) nothing is displayed.

Can you support with this?

**Comment by Snuffleupagus** [comment]
Please refer to https://github.com/mozilla/pdf.js/issues/15712#issuecomment-1332046716 specifically for the expected format of an Ink-annotation.

> I need to mimick what happens when the source PDF has a Square annotation in it, and add a new one programmatically.

Sorry, but we don't support adding Square-annotations (currently only FreeText and Ink are implemented).

> I can access the AnnotationStorage of a page, but if I push new annotations there (i.e. a copy of the original annotation from the source PDF with different coordinates) nothing is displayed.

That won't work unfortunately, since the `AnnotationStorage`-format is custom (to the PDF.js library) and consequently you simply can't copy annotation-data like that.

**Comment by ghepardo82** [comment]
Hello @Snuffleupagus ,
I am trying to do something very similar: add an annotation programmatically and be able to see it in the PDF.

Following the suggestions in the issue https://github.com/mozilla/pdf.js/issues/15712#issuecomment-1332046716 I was able to add a textAnnotation and an inkAnnotation to the storage layer, but it is not displayed.

My code looks like this:
`
    let textAnnotation1= {
          "annotationType": 3,
          "color": [1, 0, 0],
          "fontSize": 18,
          "value": "Hello World EMANUEKE JKASHDAJKHDASJKHDASKJHDASKJHD",
          "pageIndex": 0,
          "rect": [67.5, 543, 119, 556.5],
          "rotation": 0
        };
        let inkAnnotation1 = {
          "annotationType": 15,
          "color": [255, 0, 0],
          "thickness": 3,
          "opacity": 1,
          "paths": [{
            "bezier": [
              1.5, 25.727771084724367, 2.8040804485100495, 27.031851533234402,
              5.396811581133676, 23.25556095123241, 6, 22.727771084724367,
              10.45407020558315, 18.830459654839103, 15.981183968598401,
              16.364531104350363, 21, 13.227771084724367, 25.88795894206055,
              10.172796745936523, 37.988543516372076, 5.739227568352277, 42,
              1.7277710847243668
            ],
            "points": [
              1.5, 25.727771084724367, 5.225791198862495, 23.602568747729173,
              4.012834511116397, 24.914722452856147, 6, 22.727771084724367, 21,
              13.227771084724367, 37.71378602219673, 4.78737352236285,
              31.828688421912233, 7.836451889039392, 42, 1.7277710847243668
            ]
          }],
          "pageIndex": 0,
          "rect": [71.5, 534.5, 115, 562],
          "rotation": 0
        };

        let as = PDFViewerApplication.pdfDocument.annotationStorage;
        as.setValue("text_ann_1", textAnnotation1);
        as.setValue("ink_ann_2", inkAnnotation1);
`

the code works, but nothing is displayed, nor the annotation are really present in the PDF (if I download the PDF using pDFViewerApplication.save() no annotation is present).

What am I missing ?

thanks in advance for your help

**Comment by ghepardo82** [comment]
hello @Snuffleupagus , do you have any suggestions ?

Thanks

**Comment by Snuffleupagus** [comment]
You need to actually follow https://github.com/mozilla/pdf.js/issues/15712#issuecomment-1332046716 *exactly*, since the names cannot be arbitrary but must follow the demonstrated format. If you actually name your annotations correctly, then e.g. `saveDocument` should work.

However, note that manually adding annotations to the `AnnotationStorage` like this won't render them in the viewer (since that's not really a use-case that matter for e.g. the *built-in* Firefox PDF Viewer).
A possible work-around could be to change the `annotationMode` preference/option in the viewer to `3`, see [the values here](https://github.com/mozilla/pdf.js/blob/9db450966498123e23b816168de1626bf522afe1/src/shared/util.js#L60-L65), and then re-render the pages in the viewer. (Note that changing that option will disable forms, but again this particular situation isn't really one that we officially support here.)

**Comment by ghepardo82** [comment]
Thanks a lot @Snuffleupagus , indeed I noticed the requirement and the filter for the annotation names following the internal format.
I'll try to change the annotationMode of the viewer to 3 and let you know.

**Comment by ghepardo82** [comment]
I tried, but it is not rendering unfortunately :(

**Comment by rogercorrea** [comment]
Please,

I need save annotations per user because if he make annotations, I need save and show when he open again.
The resource is necessary because in N cenarious this perspective happen. I don't change the PDF file but users have personal annotations.

In a old version of the Pdf.JS I'm used a plugin (fork project) for it: https://github.com/rogercorrea/pdf-annotate.js
I will need try verify if actual PdfJS version is compatible.

**Comment by sinatayebati** [comment]
If anyone is still seeking a solution to programmatically add annotations to PDFs and use the PDF.js viewer for rendering, I highly recommend checking out this repository: [pdfAnnotate](https://github.com/highkite/pdfAnnotate).

I experienced the same pain when needed to add annotations programmatically on the server side and display them on the client side, and this incredible work saved me.

For me, using the "AnnotationStorage" feature from PDF.js proved to be complex and nearly impossible to achieve. However, after discovering this repository, I found that it provided a straightforward solution. Its integration with PDF.js is seamless, making it incredibly easy to work with. Additionally, it offers a variety of annotation types and is extensively documented, making the entire process smooth and efficient.
